ingestion-beam/src/main/java/com/mozilla/telemetry/transforms/LimitPayloadSize.java

```diff
@@ -3,47 +3,51 @@
+import org.apache.beam.sdk.values.TypeDescriptor;
 
-public class LimitPayloadSize extends MapElementsWithErrors.ToPubsubMessageFrom<PubsubMessage> {
+public class LimitPayloadSize {
 
   /** Exception to throw for messages whose payloads exceed the configured size limit. */
-  public static class PayloadTooLargeException extends Exception {
+  public static class PayloadTooLargeException extends RuntimeException {
 
     public PayloadTooLargeException(String message) {
       super(message);
     }
   }
 
   /** Return a transform that will send messages to error output if payload exceeds maxBytes. */
-  public static LimitPayloadSize toBytes(int maxBytes) {
-    return new LimitPayloadSize(maxBytes);
+  public static MapWithFailures<PubsubMessage, PubsubMessage, PubsubMessage> toBytes(int maxBytes) {
+    return LimitPayloadSize.of(maxBytes);
   }
 
   /** Return a transform that will send messages to error output if payload exceeds mb megabytes. */
-  public static LimitPayloadSize toMB(int mb) {
-    return new LimitPayloadSize(mb * 1024 * 1024);
+  public static MapWithFailures<PubsubMessage, PubsubMessage, PubsubMessage> toMB(int mb) {
+    return LimitPayloadSize.of(mb * 1024 * 1024);
   }
 
-  @Override
-  protected PubsubMessage processElement(PubsubMessage message) throws Exception {
-    message = PubsubConstraints.ensureNonNull(message);
-    int numBytes = message.getPayload().length;
-    if (numBytes > maxBytes) {
-      countPayloadTooLarge.inc();
-      throw new PayloadTooLargeException("Message payload is " + numBytes + "bytes, larger than the"
-          + " configured limit of " + maxBytes);
-    }
-    return message;
-  }
-
-  ////////
-
-  private final int maxBytes;
-
-  private final Counter countPayloadTooLarge = Metrics.counter(LimitPayloadSize.class,
-      "payload_too_large");
-
-  private LimitPayloadSize(int maxBytes) {
-    this.maxBytes = maxBytes;
+  /** Factory method to create mapper instance. */
+  public static MapWithFailures<PubsubMessage, PubsubMessage, PubsubMessage> of(int maxBytes) {
+    final Counter countPayloadTooLarge = Metrics.counter(LimitPayloadSize.class,
+        "payload_too_large");
+    return MapElements.into(TypeDescriptor.of(PubsubMessage.class)).via((PubsubMessage msg) -> {
+      msg = PubsubConstraints.ensureNonNull(msg);
+      int numBytes = msg.getPayload().length;
+      if (numBytes > maxBytes) {
+        countPayloadTooLarge.inc();
+        throw new PayloadTooLargeException("Message payload is " + numBytes
+            + "bytes, larger than the" + " configured limit of " + maxBytes);
+      }
+      return msg;
+    }).exceptionsInto(TypeDescriptor.of(PubsubMessage.class))
+        .exceptionsVia((ExceptionElement<PubsubMessage> ee) -> {
+          try {
+            throw ee.exception();
+          } catch (PayloadTooLargeException e) {
+            return FailureMessage.of("LimitPayloadSize", ee.element(), ee.exception());
```

Comment:
```suggestion
            return FailureMessage.of(LimitPayloadSize.getClass().getSimpleName(), ee.element(), ee.exception());
```

I have a slight preference for accessing the class name this way rather than repeating it as a string literal. Not sure if these are exactly the right method names.

Reply:
And to be clear, I think it makes sense to use string literals when we call `.apply(...)` in the main pipeline definitions. It's just within the context of this class itself that I'd like to avoid repeating the name.